src/protocol/serialize_stream.rs

```diff
@@ -1,426 +1,324 @@
-    let test = "test";
-
-    let mut page = MemoryPage::new(2 * 4 + arr.len() + test.len() + 
-                                   1 + 4 + 
-                                   2 + 8 + 
-                                   8);
-    let mut stream = SerializeStream::new(&mut page, 0);
-
-    stream.write_bool(true).expect("Should not fail");
-    stream.write_float(2.2).expect("Should not fail");
-    stream.write_int(2).expect("Should not fail");
-    stream.write_smallint(2).expect("Should not fail");
-    stream.write_bigint(2).expect("Should not fail");
-    stream.write_varbinary(&arr).expect("Should not fail");
-    stream.write_varchar(test).expect("Should not fail");
-    stream.write_varbinary(&[2, 2]).unwrap_err();
-}
-#[test]
-fn write_read_single_int() {
-    let mut page = MemoryPage::new(4);
-    {
-        let mut ws = SerializeStream::new(&mut page, 0);
-        ws.write_int(42).expect("Should not fail");
-    }
-
-    let mut rs = DeserializeStream::new(&mut page, 0);
-    let val = rs.read_int().unwrap();
-
-    assert_eq!(val, 42);
-}
-
-#[test]
-fn write_read_several_ints() {
-    let mut page = MemoryPage::new(4 + 4 + 4 + 3);
-    {
-        let mut ws = SerializeStream::new(&mut page, 0);
-
-        ws.write_int(2093608745).expect("Should not fail");
-        ws.write_int(0).expect("Should not fail");
-        ws.write_int(-1294).expect("Should not fail");
-    }
-
-    let mut rs = DeserializeStream::new(&mut page, 0);
-    assert_eq!(rs.read_int().unwrap(), 2093608745);
-    assert_eq!(rs.read_int().unwrap(), 0);
-    assert_eq!(rs.read_int().unwrap(), -1294);
-}
+use data_type::DataType;
+use storage::MemoryPage;
+use protocol::packer;
+
+/// Serializes and writes values to MemoryPage.
+/// Always writes values in the architertural endian currently.
+#[derive(Debug)]
+pub struct SerializeStream<'a> {
+    position: usize,
+    page: &'a mut MemoryPage,
+}
+
+impl<'a> SerializeStream<'a> {
+    /// Creates new stream.
+    pub fn new(page: &'a mut MemoryPage, pos: usize) -> Self {
+        SerializeStream {
+            position: pos,
+            page: page,
+        }
+    }
+
+    fn check_available_space(&self, val: isize) -> Result<(), String> {
+        let avail: isize = self.page.data().len() as isize - self.position as isize;
+
+        if avail < val as isize {
+            Err(format!("Not enough memory to serialize value: required={}, available={}",
+                        val,
+                        avail))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Check if we have enough memory in page to write a static typed value.
+    fn check_static_type_len(&self, field_type: DataType) -> Result<(), String> {
+        let field_len: isize = field_type.static_len() as isize;
+
+        match self.check_available_space(field_len) {
+            Err(e) => return Err(format!("Unable to write {:?}: {:?}", field_type, e)),
+            Ok(_) => ()
+        }
+
+        Ok(())

```

Comment:
Remote `Ok(())` at the end and let the `match` be a return statement here.

Reply:
Like to add Ok(_) => OK(()) ?

Reply:
Yeah, and remove `return` before `Err` part.